components/script/dom/window.rs

```diff
@@ -319,10 +319,15 @@ impl Window {
         &self.image_cache_thread
     }
 
+    /// Retrieve unsafely the browsing context
     pub fn browsing_context(&self) -> Root<BrowsingContext> {
         self.browsing_context.get().unwrap()
     }
 
+    pub fn safe_browsing_context(&self) -> Option<Root<BrowsingContext>> {
```

Comment:
The problem is that `document.has_browsing_context` is immutable, so if a document originally had a browsing context, but then that browsing context was discarded, we don't update the document. We could make `document.has_browsing_context` mutable, or we could just remove it and replace it by a method which checks to see if the window exists and has a browsing context.